internal/ls/completions.go

```diff
@@ -0,0 +1,2306 @@
+					if !isTypeLocation && !insideJsDocTagTypeExpression &&
+						core.Some(
+							symbol.Declarations,
+							func(decl *ast.Declaration) bool {
+								return decl.Kind != ast.KindSourceFile && decl.Kind != ast.KindModuleDeclaration && decl.Kind != ast.KindEnumDeclaration
+							}) {
+						t := typeChecker.GetNonOptionalType(typeChecker.GetTypeOfSymbolAtLocation(symbol, node))
+						insertQuestionDot := false
+						if typeChecker.IsNullableType(t) {
+							canCorrectToQuestionDot := isRightOfDot && !isRightOfQuestionDot &&
+								!ptrIsFalse(preferences.IncludeAutomaticOptionalChainCompletions)
+							if canCorrectToQuestionDot || isRightOfQuestionDot {
+								t = typeChecker.GetNonNullableType(t)
+								if canCorrectToQuestionDot {
+									insertQuestionDot = true
+								}
+							}
+						}
+						addTypeProperties(t, node.Flags&ast.NodeFlagsAwaitContext != 0, insertQuestionDot)
+					}
+
+					return
+				}
+			}
+		}
+
+		if !isTypeLocation || checker.IsInTypeQuery(node) {
+			// GH#39946. Pulling on the type of a node inside of a function with a contextual `this` parameter can result in a circularity
+			// if the `node` is part of the exprssion of a `yield` or `return`. This circularity doesn't exist at compile time because
+			// we will check (and cache) the type of `this` *before* checking the type of the node.
+			typeChecker.TryGetThisTypeAtEx(node, false /*includeGlobalThis*/, nil)
+			t := typeChecker.GetNonOptionalType(typeChecker.GetTypeAtLocation(node))
+
+			if !isTypeLocation {
+				insertQuestionDot := false
+				if typeChecker.IsNullableType(t) {
+					canCorrectToQuestionDot := isRightOfDot && !isRightOfQuestionDot &&
+						!ptrIsFalse(preferences.IncludeAutomaticOptionalChainCompletions)
+
+					if canCorrectToQuestionDot || isRightOfQuestionDot {
+						t = typeChecker.GetNonNullableType(t)
+						if canCorrectToQuestionDot {
+							insertQuestionDot = true
+						}
+					}
+				}
+				addTypeProperties(t, node.Flags&ast.NodeFlagsAwaitContext != 0, insertQuestionDot)
+			} else {
+				addTypeProperties(typeChecker.GetNonNullableType(t), false /*insertAwait*/, false /*insertQuestionDot*/)
+			}
+		}
+	}
+
+	if isRightOfDot || isRightOfQuestionDot {
+		getTypeScriptMemberSymbols()
+	} else if isRightOfOpenTag {
+		// !!! jsx completions
+	} else if isStartingCloseTag {
+		// !!! jsx completions
+	} else {
+		// For JavaScript or TypeScript, if we're not after a dot, then just try to get the
+		// global symbols in scope.  These results should be valid for either language as
+		// the set of symbols that can be referenced from this location.
+		// !!! global completions
+	}
+
+	var contextualType *checker.Type
+	if previousToken != nil {
+		contextualType = getContextualType(previousToken, position, file, typeChecker)
+	}
+
+	// exclude literal suggestions after <input type="text" [||] /> (#51667) and after closing quote (#52675)
```

Comment:
You do not need to do this, but it might be interesting to write out `microsoft/TypeScript#51667` or similar just to make the linking work again? Very optional, probably doesn't matter since we'll eventually move this code.